SimPEG/NSEM/ProblemNSEM.py

```diff
@@ -0,0 +1,560 @@
+from SimPEG.EM.Utils.EMUtils import omega, mu_0
+from SimPEG import SolverLU as SimpegSolver, PropMaps, Utils, mkvc, sp, np
+from SimPEG.EM.FDEM.ProblemFDEM import BaseFDEMProblem
+from SurveyNSEM import Survey, Data
+from FieldsNSEM import BaseNSEMFields, Fields1D_ePrimSec, Fields3D_ePrimSec
+from SimPEG.NSEM.Utils.MT1Danalytic import getEHfields
+import time, sys
+
+class BaseNSEMProblem(BaseFDEMProblem):
+    """
+        Base class for all Natural source problems.
+    """
+
+    def __init__(self, mesh, **kwargs):
+        BaseFDEMProblem.__init__(self, mesh, **kwargs)
+        Utils.setKwargs(self, **kwargs)
+    # Set the default pairs of the problem
+    surveyPair = Survey
+    dataPair = Data
+    fieldsPair = BaseNSEMFields
+
+    # Set the solver
+    Solver = SimpegSolver
+    solverOpts = {}
+
+    verbose = False
+    # Notes:
+    # Use the forward and devs from BaseFDEMProblem
+    # Might need to add more stuff here.
+
+    ## NEED to clean up the Jvec and Jtvec to use Zero and Identities for None components.
+    def Jvec(self, m, v, f=None):
+        """
+        Function to calculate the data sensitivities dD/dm times a vector.
+
+            :param numpy.ndarray m (nC, 1) - conductive model
+            :param numpy.ndarray v (nC, 1) - random vector
+            :param NSEMfields object (optional) - NSEM fields object, if not given it is calculated
+            :rtype: NSEMdata object
+            :return: Data sensitivities wrt m
+        """
+
+        # Calculate the fields
+        if f is None:
+           f= self.fields(m)
+        # Set current model
+        self.curModel = m
+        # Initiate the Jv object
+        Jv = self.dataPair(self.survey)
+
+        # Loop all the frequenies
+        for freq in self.survey.freqs:
+            dA_du = self.getA(freq) #
+
+            dA_duI = self.Solver(dA_du, **self.solverOpts)
+
+            for src in self.survey.getSrcByFreq(freq):
+                # We need fDeriv_m = df/du*du/dm + df/dm
+                # Construct du/dm, it requires a solve
+                # NOTE: need to account for the 2 polarizations in the derivatives.
+                u_src = f[src,:] # u should be a vector by definition. Need to fix this...
+                # dA_dm and dRHS_dm should be of size nE,2, so that we can multiply by dA_duI. The 2 columns are each of the polarizations.
```

Comment:
This is longer-term thinking / rambling, so no need to address it in this pr. 

I would suggest that the handling of the two polarizations be done in the final step - with the derivatives on the receivers. (represent the two polarizations as two individual sources, recognizing that the receiver needs information from two sources to evaluate itself). This would require some re-thinking on how receivers and sources are defined - I would perhaps see them as being defined independently, then "wiring" put in place (in the definition of the survey) to connect sources, receivers together in a meaningful manner. 

Perhaps something along the lines of 
```python
src_px = NSEM.Src.PlaneWave_x(freq=f) # and we might define src = NSEM.src.planeWave(freq=f) that is a composite source, and contains both polarizations   
src_py = NSEM.Src.PlaneWave_y(freq=f)
zxy = NSEM.Rx.Impedance_xy(locs)
zyx = NSEM.Rx.Impedance_yx(locs)
survey = NSEM.survey([([src_px, src_py], [zxy, zyx])]) # eg. a list of tuples (src, rx ) or ([srcs], [rxs])
```

Thoughts @rowanc1, @grosenkj ?